### DISCUSSION TITLE
ローカルレコードがオンでアプリがバックグラウンドからフォアグラウンドに復帰した際は引き続きローカルレコードはできますか？

### DISCUSSION THREAD

**Discussion by yuya-h-29** [HaishinKit/HaishinKit.swift#1006] [Q&A] [answered]
**Describe the bug**
バグというよりも質問になってしまいます。すみません。
ローカルレコードを選択状態で配信し、アプリがバックグランドに入ってからフォアグラウンドに戻った際は

- 続きからローカルレコードを再開して1つの動画としてカメラロールに保存をする
- （別の動画ファイルに分けられてしまってもいいので）配信中の動画をすべてカメラロールに保存する

ことはできますか？

#412 と #612 あたりを見ながら実装し、配信後アプリが常にフォアグラウンドに居続ければローカルレコードは成功するのですが、一度バックグラウンドに入ってからの処理がわからず、フォアグラウンドに復帰したあとの動画すべてをローカルレコードすることができていません。

**To Reproduce**
1. （正しい実装方法がわかりませんんが）バックグラウンドに入ったら`rtmpStream.close()`を呼ぶ
2. `didFinishWriting`が呼ばれ、それまでの配信していた動画がローカルレコードされる
3. アプリをフォアグラウンドに戻し、もう一度`rtmpStream.publish(streamKey, type: .localRecord)`を呼ぶ
5. ローカルレコードが上手くいく際はファイルのパスがコンソールに表示される`file:///var/mobile/Containers/Data/Application/.....mp4`が、ローカルレコードが上手くいかない際はファイルパスが表示されず、数秒後に`unregister notification for read_timeout failed`と`unregister notification for write_timeout failed`がコンソールに表示され、更に数秒後に`RTMPConnection.Code.connectClosed`が呼ばれる。

ローカルレコード有りの場合はアプリがバックグラウンド・フォアグラウンドに入ったタイミングでどのメソッドを使うとローカルレコードが成功しますでしょうか？

```
class LiveCameraVC: UIViewController {
    override func viewDidLoad() {
        NotificationCenter.default.addObserver(self, selector: #selector(didEnterBackground(_:)), name: UIApplication.didEnterBackgroundNotification, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(didBecomeActive(_:)), name: UIApplication.didBecomeActiveNotification, object: nil)
    }

    @objc func didEnterBackground(_ notification: Notification) {
        if enableLocalRecord {
            rtmpStream.close() // 実装方法がわからず、ローカルレコードが有りの場合はcloseしてdidFinishWritingが無理やり呼ばれるようにしています…
        } else {
            rtmpStream.receiveVideo = false
            rtmpStream.receiveAudio = false
        }
    }
    
    @objc func didBecomeActive(_ notification: Notification) {
        if enableLocalRecord {
            rtmpStream.publish(targetEvent.streamKey, type: .localRecord)
        } else {
            // ローカルレコードなしの場合はフォアグラウンドに復帰したところから再配信できる
            rtmpStream.receiveVideo = true
            rtmpStream.receiveAudio = true
        }
    }
}

class RecorderDelegate: DefaultAVRecorderDelegate {
    override func didFinishWriting(_ recorder: AVRecorder) {
        guard let writer: AVAssetWriter = recorder.writer else { return }
        PHPhotoLibrary.shared().performChanges({() -> Void in
            PHAssetChangeRequest.creationRequestForAssetFromVideo(atFileURL: writer.outputURL)
        }, completionHandler: {saved, error -> Void in
            if let _ = error {
                // do something
            } else if saved {
                // do something
            } else {
                // do something
            }
            do {
                try FileManager.default.removeItem(at: writer.outputURL)
            } catch {
                print("error in recorder delegate: \(error.localizedDescription)")
            }
        })
    }
}
```

**Expected behavior**

-　ローカルレコードを選択時は、アプリがフォアグラウンドに居続ける間はローカルレコードをしたい。
- フォアグラウンドからバックグラウンドに入り、そこまでの動画をローカルレコードし、再度フォアグラウンドに復帰した場合はそこからローカルレコードを再開したい。
- バックグラウンドとフォアグラウンドを複数回行き来してもローカルレコードでフォトライブラリアプリに保存される動画は1本にできると良いが、可能でなけれは複数の動画ファイルに別れてしまっても良い。

**Smartphone (please complete the following information):**
 - Device: iPhoneXS max
 - OS: iOS14.8.1

よろしくお願いいたします。

**Comment by shogo4405** [comment, marked as answer]
作者ですが。その要件にしたがってテストしたことがないのでわからないです。

1. https://github.com/shogo4405/HaishinKit.swift/issues/926 の設定した後だと、マイクが有効のままになるかと思います。この設定だとバックグラウンド状態でもマイクを利用することができ。アプリケーションも起動をつづけることができます。
2. この時に、バックグラウンドで継続して録画できないでしょうか？

**Comment by yuya-h-29** [comment]
@shogo4405 

コメント有難うございます。
#926 を確認したところ、すでに自分のプロジェクトにそちらのコードを入れていました。（再確認するきっかけにもなりました。有難うございます）

1点だけ教えていただきたいのですが、
`RTMPConnection` と`RTMPStream`はざっくり
`RTMPConnection` がクライアントとサーバーのコネクションを作り、`RTMPStream`が実際に配信を送るクラスのような認識で合っていますか？いままでこの辺りを曖昧で開発をしてきてしまい、もし認識が間違っていたら教えていただきたいです。

そして前に質問したバックグラウンドに入ったときのローカルレコードについてですが、最終的に問題は解決できました！
今まではバックグラウンドに入ってから、`rtmpStream.close()`を読んで配信を無理やり終了していましたが、`rtmpConnection.close()` を呼べば配信を無理やり終了しなくともバックグラウンドに入るまでの動画のローカルレコードをすることができました。
フォアグラウンドに復帰した際に`rtmpConnection.connect()`で再度接続すれば、接続後〜終了まで（または再度バックグラウンドに入るまで）の動画をローカル保存することができました。動画ファイルは別になりますが、ひとまず大丈夫そうです。


    @objc func didEnterBackground(_ notification: Notification) {
        if enableLocalRecord && isLiveStarted {
            rtmpConnection.close()
        } else {
            rtmpStream.receiveVideo = false
            rtmpStream.receiveAudio = false
        }
    }
    
    @objc func didBecomeActive(_ notification: Notification) {
        if enableLocalRecord && isLiveStarted {
            rtmpConnection.connect(targetEvent.rtmpUrl)
        } else {
            rtmpStream.receiveVideo = true
            rtmpStream.receiveAudio = true
        }
    }

**Reply by shogo4405** [reply]
はい。そちらの認識で大丈夫ですよ。
